core/pta/rtc.c

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2018, Linaro Limited
+ * Copyright (c) 2021, EPAM Systems. All rights reserved.
+ * Copyright (c) 2022, Microchip
+ *
+ * Based on pta/hwrng.c
+ *
+ */
+
+#include <drivers/rtc.h>
+#include <kernel/pseudo_ta.h>
+#include <rtc_pta_client.h>
+#include <tee_api_defines.h>
+
+#define PTA_NAME "rtc.pta"
+
+static TEE_Result rtc_pta_get_time(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct optee_rtc_time *time = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	time = params[0].memref.buffer;
+	if (!time || params[0].memref.size != sizeof(*time))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	rtc_get_time(time);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result rtc_pta_set_time(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct optee_rtc_time *time = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	time = params[0].memref.buffer;
+	if (!time || params[0].memref.size != sizeof(*time))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	rtc_set_time(time);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result rtc_pta_set_offset(uint32_t types,
+				     TEE_Param params[TEE_NUM_PARAMS])
+{
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	rtc_set_offset(params[0].value.a);
```

Comment:
Since an offset can be negative I suppose that even though the value is a `uint32_t` it should be casted into a `int32_t` in order to be properly upgraded to a `long` on 64bit platforms.